src/sentry/integrations/slack/tasks.py

```diff
@@ -229,3 +239,71 @@ def find_channel_id_for_alert_rule(organization_id, uuid, data, alert_rule_id=No
     # some other error
     redis_rule_status.set_value("failed")
     return
+
+
+@instrumented_task(name="sentry.integrations.slack.link_users_identities", queue="integrations")
+def link_slack_user_identities(integration, organization):
+    emails_by_user = get_emails_by_user(organization)
+    slack_data_by_user = get_slack_data_by_user(integration, organization, emails_by_user)
+
+    idp = IdentityProvider.objects.get(
+        type=integration.provider,
+        external_id=integration.external_id,
+    )
+    identity_data = {
+        "status": IdentityStatus.VALID,
+        "date_verified": timezone.now(),
+    }
+    identity_models = Identity.objects.filter(
+        idp=idp,
+        user__in=slack_data_by_user.keys(),
+        status=IdentityStatus.VALID,
+    )
+
+    identities_by_user = {identity.user: identity for identity in identity_models}
+    for user, data in slack_data_by_user.items():
+        if (
+            user in identities_by_user.keys()
+            and data["email"] in emails_by_user[user]
+            and data["slack_id"] != identities_by_user[user].external_id
+        ):
+            # identity already exists, the emails match, AND the external ID has changed
+            identities_by_user[user].update(external_id=data["slack_id"])
+        if user not in identities_by_user.keys() and data["email"] in emails_by_user[user]:
```

Comment:
Both of the conditionals check the same condition: `data["email"] in emails_by_user[user]`.
I think we can add this as the top of the loop:
```
if data["email"] not in emails_by_user[user]:
	continue
```